Add support for Levoit Core 300S air purifier to VeSync integration

RefineryX opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
First PR! Following the successful merge of https://github.com/home-assistant/core/pull/57126, this small pull request adds support for the Leviot Core 300s air purifier to the VeSync integration.

Currently, this integration does not expose any sensors.

The Core 300s was previously unable to be supported due to issues with pyvesync, however this has now been corrected with the following PR https://github.com/webdjoe/pyvesync/pull/94. 

As such, the dependency requires 1.4.2 https://github.com/webdjoe/pyvesync/releases/tag/1.4.2

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/20809

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

RefineryX committed:
Update fan.py

Adds support for Core 300s

RefineryX committed:
Update manifest.json

RefineryX committed:
Update requirements_all.txt

RefineryX committed:
Update requirements_test_all.txt

homeassistant commented:
Hi @RefineryX, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#62488).

Once you do that we will be able to review and accept this pull request.

Thanks!

frenck commented on homeassistant/components/vesync/fan.py:
```diff
@@ -124,6 +126,9 @@ def extra_state_attributes(self):
         if hasattr(self.smartfan, "night_light"):
             attr["night_light"] = self.smartfan.night_light
 
+        if hasattr(self.smartfan, "display_state"):
+            attr["display_state"] = self.smartfan.display_state
+
```
Shouldn't this be a separate sensor?

RefineryX replied:
Currently, this integration does not have any sensors exposed.  When adding it to home assistant the only thing that is there is 1 ‘entity’ which is the main fan. I’d love to be able to expose the main sensors but fear it’s above my skill level. Lines 129-131 add to the already existing attributes. I can remove this line from this PR if needed. 

For reference, here are the state attributes for this entity.  

```
speed_list:
  - 'off'
  - low
  - medium
  - high
  - auto
  - sleep
preset_modes:
  - auto
  - sleep
speed: low
percentage: 33
percentage_step: 33.333333333333336
preset_mode: null
child_lock: false
night_light: 'off'
air_quality: 6
mode: manual
filter_life: 60
display_state: false
friendly_name: Bedroom Purifier
supported_features: 1 
```

frenck replied:
> Currently, this integration does not have any sensors exposed.

Then we should add that. We should not introduce more items of a pattern we actually no longer accept in Home Assistant.

> I’d love to be able to expose the main sensors but fear it’s above my skill level. Lines 129-131 add to the already existing attributes. I can remove this line from this PR if needed.

In that case, removing the new attribute sounds like an option.

RefineryX replied:
Thank you. I have removed that from the file - hoping it has filtered through to this request. Please let me know if there is anything else. I will begin researching adding sensors and hope to maybe add that via another PR later this year.

RefineryX committed:
Update fan.py

Removed request to include attribute for 'display_state' given that HA no longer accepts.

frenck reviewed (approved):
Thanks, @RefineryX 👍

decompil3d commented:
PR #61837 adds VeSync devices to the device registry and introduces sensors for power/energy use on outlets. So after that's merged you could build a follow-up PR that uses the new `sensor.py` to make sensors for your fan attributes as well.
